kernel/src/metaptr.rs

```diff
@@ -0,0 +1,113 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Google LLC 2024.
+
+//! Defines the MetaPtr type
+
+use core::fmt::{Formatter, LowerHex, UpperHex};
+use core::ops::AddAssign;
+
+/// A pointer with target specific metadata.
+/// This should be used any time the kernel wishes to grant authority to the user, or any time
+/// the user should be required to prove validity of a pointer.
```

Comment:
I wonder about the second bullet point. On CHERI specifically, the register size happens to be the meta-dataish pointer rather than usize. Maybe, more specifically an integer that should fix exactly in a register?

I would also say that the distinction between where to use `usize` vs `* const T` is that the first is just an address or size (but is not meant to imply the existence of an object), while the second implies there is one or more T's there and code may make some attempt to dereference it (I think either in the kernel or userspace).

I will have another go at the text.

Reply:
> I wonder about the second bullet point. On CHERI specifically, the register size happens to be the meta-dataish pointer rather than usize. Maybe, more specifically an integer that should fix exactly in a register?

Right, the actual point (of this point!) is that there are cases where we want to say something about, basically, how arguments to a function will be packed into registers.

For example, the arguments to `command` are arbitrary values that each fit into a register (individual drivers then interpret those values as some specific type, such as a `u32`).

Another example, the process struct needs to store registers from a context switch. Some of those are pointers of some kind (`pc`, `sp`, etc), but most are general purpose registers, and their "type" is `usize`---it's whatever size the hardware uses for registers (the meaning of the value could be a size, it could two i16s, it could be anything...) and also it is not a to be interpreted by the kernel as a pointer.

Perhaps this is more complicated because actually in CHERI, the process is allowed to store a capability (which is larger than a `usize`?) in a general purpose register and the kernel _needs_ to actually store that in the process struct?

> I would also say that the distinction between where to use usize vs * const T is that the first is just an address or size (but is not meant to imply the existence of an object), while the second implies there is one or more T's there and code may make some attempt to dereference it (I think either in the kernel or userspace).

Yes, I agree with this.

Reply:
> For example, the arguments to `command` are arbitrary values that each fit into a register (individual drivers then interpret those values as some specific type, such as a `u32`).

I couldn't really find a good answer to this in the syscalls TRD, but don't we have to be more specific than this? Like if the platform was some hypothetical 31 bit hardware we would still need to provide command with 32 bit arguments, or if we had a 64 bit platform, we don't want capsules that need 64 bits per command arg, do we?

Reply:
Oh good point. It's possible `command` is a bad example here if we actually _do_ want `command` to take N arguments of 32 bits each and on, e.g., a 64 bit platform those might be packed two to a register (say, if that's the calling convention).